### PR TITLE
feat(trace): implement tree representation

### PR DESCRIPTION
**Pull request by JonasBa**
Implements the tree representation of the trace view with a flattened list iterator. This has backend feature parity with current view and allows us to extend and mutate the tree in the future to handle incremental rendering. 

There is a large code comment inside the tracetree file that explains how the code works and behaves. Relevant tests were added for the tree structure, not for the UI as that is still much of a wip

### CODE REVIEW

File: static/app/views/performance/newTraceDetails/guards.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    node.value &&
+    'type' in node.value &&
+    node.value.type === 'missing_instrumentation'
+  );
+}
+
+export function isSpanNode(
+  node: TraceTreeNode<TraceTree.NodeValue>
+): node is TraceTreeNode<TraceTree.Span> {
+  return !!(node.value && ('description' in node.value || 'span_id' in node.value));
```
Comment (Abdkhan14) [inline static/app/views/performance/newTraceDetails/guards.tsx:21]: Transactions have span_ids too, but no description. we should just stick to `'description' in node.value`

Comment (JonasBa) [reply static/app/views/performance/newTraceDetails/guards.tsx:21]: Spans cant have description field right?

**Review by Abdkhan14 — APPROVED** [review]
Hey we paired through all of this code and did rigorous manual testing:
- We need comments especially in `traceTree.tsx` for the tree building logic.
- We added Todo comments for blocks that can be improved. 
- The code is behind the feature flag `organizations:trace-view-v1`
- We will move render blocks for different row typesin the traceview to there own component files.
This can all be done in following PRs. Lets merge this beast.

File: static/app/views/performance/newTraceDetails/guards.tsx (hunk 1 above)
Comment (Abdkhan14) [reply static/app/views/performance/newTraceDetails/guards.tsx:21]: transactions can't, all spans do.

**Comment by JonasBa** [comment]
@Abdkhan14 I see some spans without a description :)
CleanShot 2024-02-09 at 15 23 50@2x

**Pull request merged**